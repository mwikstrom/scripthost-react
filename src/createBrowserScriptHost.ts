import { ScriptHost, ScriptHostOptions } from "scripthost";
import { BrowserSandbox, BrowserSandboxOptions } from "scripthost-browser";

/**
 * @public
 */
export function createBrowserScriptHost(
    options: ScriptHostOptions & Pick<BrowserSandboxOptions, "unsafe"> = {}
): ScriptHost {
    const { unsafe, ...hostOptions } = options;
    return new ScriptHost(() => new BrowserSandbox({ unsafe }), { ...hostOptions });
}
