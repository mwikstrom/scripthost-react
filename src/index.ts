/**
 * React bindings for Scripthost
 * @packageDocumentation
 */

export * from "./createBrowserScriptHost";
export * from "./ScriptHostScope";
export * from "./useObservedScript";
export * from "./useScriptInvoker";
