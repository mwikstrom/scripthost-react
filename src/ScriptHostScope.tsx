import React, { createContext, FC, ReactNode, useContext, useMemo } from "react";
import { ScriptHost } from "scripthost";
import { createBrowserScriptHost } from "./createBrowserScriptHost";

/**
 * @public
 */
export interface ScriptHostScopeProps {
    children?: ReactNode;
    host?: ScriptHost;
}

/**
 * @public
 */
export const ScriptHostScope: FC<ScriptHostScopeProps> = props => {
    const { children, host } = props;
    const getter = useMemo(() => host ? () => host : once(createBrowserScriptHost), [host]);
    return (
        <ScriptHostContext.Provider
            value={getter}
            children={children}
        />
    );
};

/**
 * @public
 */
export const useScriptHost = (): ScriptHost => useContext(ScriptHostContext)();

const once = <T,>(func: () => T): () => T => {
    let memo: T | undefined;
    return () => {
        if (memo === undefined) {
            memo = func();
        }
        return memo;
    };
};

type ScriptHostGetter = () => ScriptHost;
const ScriptHostContext = createContext<ScriptHostGetter>(once(createBrowserScriptHost));
