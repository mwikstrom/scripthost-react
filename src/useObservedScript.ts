import { useEffect, useMemo, useState } from "react";
import { ScriptEvalOptions, ScriptHost, ScriptObserveOptions } from "scripthost";
import { ScriptValue } from "scripthost-core";
import { isNonVoidScript } from "./internal/void-script";
import { useScriptHost } from "./ScriptHostScope";

/**
 * @public
 */
export interface ObservedScript {
    result: ScriptValue;
    ready: boolean;
    error: Error | null;
}

/**
 * @public
 */
export type UseObservedScriptOptions = Pick<ScriptEvalOptions, "instanceId">;

/**
 * @public
 */
export function useObservedScript(script: string | null, options: UseObservedScriptOptions = {}): ObservedScript {
    const { instanceId } = options;
    const host = useScriptHost();
    const entry = useMemo(() => {
        if (isNonVoidScript(script)) {
            return getCacheEntry(host, script, instanceId);
        } else {
            return null;
        }
    }, [host, script, instanceId]);
    const [output, setOutput] = useState(entry ? entry.output : successOutput(void(0)));

    useEffect(() => {
        if (entry) {
            setOutput(entry.output);
            return entry.observe(setOutput);
        }
    }, [entry]);

    return output;
}

class CacheEntry {
    readonly #callbacks = new Map<(output: ObservedScript) => void, number>();
    readonly #host: ScriptHost;
    readonly #script: string;
    readonly #instanceId: string | undefined;
    #output = initialOutput;
    #dispose: (() => void) | null = null;

    constructor(host: ScriptHost, script: string, instanceId: string | undefined) {
        this.#host = host;
        this.#script = script;
        this.#instanceId = instanceId;
    }

    get output(): ObservedScript {
        return this.#output;
    }

    get active(): boolean {
        return this.#dispose !== null;
    }

    observe(callback: (output: ObservedScript) => void): () => void {
        let active = true;
        const incremented = (this.#callbacks.get(callback) ?? 0) + 1;
        this.#callbacks.set(callback, incremented);

        if (!this.active) {
            this.#start();
        }

        return () => {
            if (active) {
                const decremented = (this.#callbacks.get(callback) ?? 0) - 1;
                active = false;
                if (decremented <= 0) {
                    this.#callbacks.delete(callback);
                    if (this.#callbacks.size === 0) {
                        setTimeout(() => {
                            if (this.active && this.#callbacks.size === 0) {
                                this.#stop();
                            }
                        }, LINGER_TIME_MS);
                    }
                }
            }
        };
    }

    #start() {
        if (this.#dispose) {
            return;
        }

        const options: ScriptObserveOptions = {
            instanceId: this.#instanceId,
            onNext: result => this.#onNext(result),
            onError: error => this.#onError(error),
        };

        this.#dispose = this.#host.observe(this.#script, options);
    }

    #stop() {
        if (!this.#dispose) {
            return;
        }

        try {
            this.#dispose();
        } catch (error) {
            console.error("Disposal of observed script threw exception:", error);
        }

        this.#dispose = null;
    }

    #onNext(result: ScriptValue) {
        this.#output = successOutput(result);
        this.#notify();
    }

    #onError(error: unknown) {
        this.#output = errorOutput(error instanceof Error ? error : new Error("Script failed"));
        this.#notify();
    }

    #notify() {
        for (const callback of this.#callbacks.keys()) {
            try {
                callback(this.#output);
            } catch (error) {
                console.error("Observed script callback threw exception:", error);
            }
        }
    }
}

const getCacheEntry = (host: ScriptHost, script: string, instanceId: string | undefined): CacheEntry => {
    let perHost = PER_HOST_CACHE.get(host);
    if (!perHost) {
        PER_HOST_CACHE.set(host, perHost = new Map());
    }

    let perScript = perHost.get(script);
    if (!perScript) {
        perHost.set(script, perScript = new Map());
    }

    let perInstance = perScript.get(instanceId);
    if (!perInstance) {
        perScript.set(instanceId, perInstance = new CacheEntry(host, script, instanceId));
    }

    return perInstance;
};

const PER_HOST_CACHE = new WeakMap<ScriptHost, Map<string, Map<string | undefined, CacheEntry>>>();

const initialOutput: ObservedScript = Object.freeze({
    result: undefined,
    ready: false,
    error: null,
});

const successOutput = (result: ScriptValue): ObservedScript => Object.freeze({
    result,
    ready: true,
    error: null,
});

const errorOutput = (error: Error): ObservedScript => Object.freeze({
    result: undefined,
    ready: true,
    error,
});

const LINGER_TIME_MS = 1000;