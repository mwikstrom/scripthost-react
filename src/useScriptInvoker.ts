import { useCallback } from "react";
import { ScriptEvalOptions } from "scripthost";
import { ScriptValue } from "scripthost-core";
import { isNonVoidScript } from "./internal/void-script";
import { useScriptHost } from "./ScriptHostScope";

/**
 * @public
 */
export type ScriptInvoker = () => Promise<ScriptValue>;

/**
 * @public
 */
export type UseScriptInvokerOptions = Pick<ScriptEvalOptions, "instanceId" | "timeout" | "vars">;

/**
 * @public
 */
export function useScriptInvoker(script: string | null, options: UseScriptInvokerOptions = {}): ScriptInvoker {
    const { instanceId, timeout, vars } = options;
    const host = useScriptHost();
    return useCallback<ScriptInvoker>(async () => {
        if (isNonVoidScript(script)) {
            await host.eval(script, { instanceId, timeout, vars });
        } else {
            return void(0);
        }
    }, [host, script, instanceId, timeout, vars]);
}