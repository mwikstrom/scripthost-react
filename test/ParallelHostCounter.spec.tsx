/* eslint-disable @typescript-eslint/no-non-null-assertion */
import React, { FC } from "react";
import { create, act, ReactTestRenderer } from "react-test-renderer";
import { ScriptHost } from "scripthost";
import { createBrowserScriptHost, ScriptHostScope, useObservedScript, useScriptInvoker } from "../src";

test("Parallel host counter", async () => {
    const host1 = createBrowserScriptHost({ unsafe: true });
    const host2 = createBrowserScriptHost({ unsafe: true });
    const element1 = <CounterControl host={host1}/>;
    const element2 = <CounterControl host={host2}/>;
    
    // Initialzing
    let renderer1: ReactTestRenderer;
    let renderer2: ReactTestRenderer;
    act(() => {
        renderer1 = create(element1);
        renderer2 = create(element2);
    });
    expect(renderer1!.toJSON()).toMatchSnapshot("[1] Initializing");
    expect(renderer2!.toJSON()).toMatchSnapshot("[2] Initializing");
    
    // Expression evaluated => Value = 0
    await act(async () => {
        await host1.whenIdle();
        await host2.whenIdle();
        renderer1.update(element1);
        renderer2.update(element2);
    });
    expect(renderer1!.toJSON()).toMatchSnapshot("[1] Value = 0");
    expect(renderer2!.toJSON()).toMatchSnapshot("[2] Value = 0");

    // Click increment button => Value = 1
    await act(async () => {
        renderer1.root.findByType("button").props.onClick();
        renderer2.root.findByType("button").props.onClick();
        await host1.whenIdle();
        await host2.whenIdle();
        renderer1.update(element1);
        renderer2.update(element2);
    });
    expect(renderer1!.toJSON()).toMatchSnapshot("[1] Value = 1");
    expect(renderer2!.toJSON()).toMatchSnapshot("[2] Value = 1");

    // Click increment button => Value = 2
    await act(async () => {
        renderer1.root.findByType("button").props.onClick();
        renderer2.root.findByType("button").props.onClick();
        await host1.whenIdle();
        await host2.whenIdle();
        renderer1.update(element1);
        renderer2.update(element2);
    });
    expect(renderer1!.toJSON()).toMatchSnapshot("[1] Value = 2");
    expect(renderer2!.toJSON()).toMatchSnapshot("[2] Value = 2");

    // Unmount and wait for observed script linger timeout
    jest.useFakeTimers();
    act(() => {
        renderer1!.unmount();
        renderer2!.unmount();
    });
    jest.runAllTimers();
});

const CounterControl: FC<{ host: ScriptHost }> = ({host}) => (
    <ScriptHostScope host={host}>
        <IncrementButton/>
        <CounterValue/>
    </ScriptHostScope>
);

const CounterValue: FC = () => {
    const { result, ready } = useObservedScript("value || 0");
    if (ready) {
        return <div>Value = {String(result)}</div>;
    } else {
        return <div>Initializing</div>;
    }
};

const IncrementButton: FC = () => {
    const increment = useScriptInvoker("value = (value || 0) + 1");
    return <button onClick={increment}>Increment</button>;
};
