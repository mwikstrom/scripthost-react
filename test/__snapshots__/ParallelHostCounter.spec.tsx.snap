// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Parallel host counter: [1] Initializing 1`] = `
Array [
  <button
    onClick={[Function]}
  >
    Increment
  </button>,
  <div>
    Initializing
  </div>,
]
`;

exports[`Parallel host counter: [1] Value = 0 1`] = `
Array [
  <button
    onClick={[Function]}
  >
    Increment
  </button>,
  <div>
    Value = 
    0
  </div>,
]
`;

exports[`Parallel host counter: [1] Value = 1 1`] = `
Array [
  <button
    onClick={[Function]}
  >
    Increment
  </button>,
  <div>
    Value = 
    1
  </div>,
]
`;

exports[`Parallel host counter: [1] Value = 2 1`] = `
Array [
  <button
    onClick={[Function]}
  >
    Increment
  </button>,
  <div>
    Value = 
    2
  </div>,
]
`;

exports[`Parallel host counter: [2] Initializing 1`] = `
Array [
  <button
    onClick={[Function]}
  >
    Increment
  </button>,
  <div>
    Initializing
  </div>,
]
`;

exports[`Parallel host counter: [2] Value = 0 1`] = `
Array [
  <button
    onClick={[Function]}
  >
    Increment
  </button>,
  <div>
    Value = 
    0
  </div>,
]
`;

exports[`Parallel host counter: [2] Value = 1 1`] = `
Array [
  <button
    onClick={[Function]}
  >
    Increment
  </button>,
  <div>
    Value = 
    1
  </div>,
]
`;

exports[`Parallel host counter: [2] Value = 2 1`] = `
Array [
  <button
    onClick={[Function]}
  >
    Increment
  </button>,
  <div>
    Value = 
    2
  </div>,
]
`;
